core-authentication/src/main/java/gg/solarmc/loader/authentication/AuthenticationCenter.java

```diff
@@ -51,13 +68,78 @@
+		DSLContext context = transaction.getProperty(DSLContext.class);
+		/*
+		 * It is acceptable to rely on the fact that any entry in the auth passwords
+		 * table will be matched by a corresponding one in the name history table.
+		 * The reason is because as soon as an account is created, the user's data
+		 * is also loaded, and their name history is catalogued, all within the
+		 * same transaction.
+		 *
+		 * This query also relies on the case insensitivity of MariaDB
+		 */
+		Record6<byte[], String, Byte, Integer, byte[], byte[]> accountRecord = context
+				.select(LATEST_NAMES.UUID, LATEST_NAMES.USERNAME,
+						AUTH_PASSWORDS.ITERATIONS, AUTH_PASSWORDS.MEMORY,
+						AUTH_PASSWORDS.PASSWORD_HASH, AUTH_PASSWORDS.PASSWORD_SALT)
+				.from(LATEST_NAMES)
+				.innerJoin(AUTH_PASSWORDS)
+				.on(AUTH_PASSWORDS.USERNAME.eq(LATEST_NAMES.USERNAME))
+				.where(LATEST_NAMES.USERNAME.eq(username))
+				.fetchOne();
+		if (accountRecord == null) {
+			return AutoLoginPreparation.noneFound();
+		}
+		UserDetails userDetails = new UserDetails(
+				UUIDUtil.fromByteArray(accountRecord.get(LATEST_NAMES.UUID)),
+				accountRecord.get(LATEST_NAMES.USERNAME));
+		if (userDetails.isPremium()) {
+			assert accountRecord.get(AUTH_PASSWORDS.ITERATIONS) == 0 : "premium accounts use iterations=0";
+			return AutoLoginPreparation.forExistingPremiumUser(userDetails);
+		}
+		if (!userDetails.username().equals(username)) {
+			return AutoLoginPreparation.deniedCaseSensitivityOfName();
+		}
+		return AutoLoginPreparation.forExistingCrackedUser(
+				userDetails,
+				new VerifiablePassword(
+						new PasswordHash(accountRecord.get(AUTH_PASSWORDS.PASSWORD_HASH)),
+						new PasswordSalt(accountRecord.get(AUTH_PASSWORDS.PASSWORD_SALT)),
+						new HashingInstructions(
+								accountRecord.get(AUTH_PASSWORDS.ITERATIONS),
+								accountRecord.get(AUTH_PASSWORDS.MEMORY))
+				));
+	}
+
+	/**
+	 * Determines whether the new user needs to enter a password. The user is deemed new on the
+	 * basis that {@link #findExistingUserWithName(Transaction, String)} does not find an existing
+	 * user. <br>
+	 * <br>
+	 *  By now, the caller should either have a premium user who has authenticated via the login protocol,
+	 * or a cracked user. If the user is premium, they may still need to enter a password if there is
+	 * a cracked account with the same name. If the user is cracked, they will need to enter a password,
+	 * but it is also possible they are denied for other reasons.
 	 *
 	 * @param transaction the transaction
 	 * @param user the unauthenticated user
 	 * @return the login attempt result
 	 */
-	public InitialLoginAttempt attemptInitialLogin(Transaction transaction, UnauthenticatedUser user) {
+	public LoginAttempt attemptLoginFromNewUsername(Transaction transaction, UserWithDataNotYetLoaded user) {
 		DSLContext context = transaction.getProperty(DSLContext.class);
 
 		Record5<String, Byte, Integer, byte[], byte[]> passwordRecord = context
```

Comment:
Is there a reason you're using skeleton Records instead of generated records?

Reply:
Could you explain what you consider a skeleton record and a generated record? Are you referring to Record5 versus something like TableNameRecord?

Reply:
Yes

Reply:
I never use the ORM-like TableNameRecord classes, because they always select every column value. Even if your query happens to select all columns at the moment, if you ever add more columns, your existing query starts to select them, unnecessarily.

Selecting unnecessary columns is something you should avoid if you want to write performant SQL
https://blog.jooq.org/2017/03/08/many-sql-performance-problems-stem-from-unnecessary-mandatory-work/